network/gateway_test.go

```diff
@@ -0,0 +1,132 @@
+			"default via 10.100.1.10 dev lxdbr0\n" +
+			"10.0.0.0/24 dev wlp2s0 proto kernel scope link src 10.0.0.66 metric 600\n" +
+			"10.100.1.0/24 dev lxdbr0 proto kernel scope link src 10.100.1.1\n", nil
+	})
+	ip, dev, err := network.GetDefaultRoute()
+	c.Assert(err, gc.IsNil)
+	c.Check(ip, gc.DeepEquals, net.ParseIP("10.100.1.10"))
+	c.Check(dev, gc.Equals, "lxdbr0")
+}
+
+func (s *GatewaySuite) TestDefaultRouteLinuxNoGW(c *gc.C) {
+	s.PatchValue(network.SimulatedOS, "linux")
+	s.PatchValue(network.LaunchIpRouteShow, func() (string, error) {
+		return "default via 10.0.0.1 dev wlp2s0 proto static metric 800\n" +
+			"default dev lxdbr0\n" +
+			"10.0.0.0/24 dev wlp2s0 proto kernel scope link src 10.0.0.66 metric 600\n" +
+			"10.100.1.0/24 dev lxdbr0 proto kernel scope link src 10.100.1.1\n", nil
+	})
+	ip, dev, err := network.GetDefaultRoute()
+	c.Assert(err, gc.IsNil)
+	c.Check(ip, gc.IsNil)
+	c.Check(dev, gc.Equals, "lxdbr0")
+}
+
+func (s *GatewaySuite) TestDefaultRouteLinuxNoDev(c *gc.C) {
+	s.PatchValue(network.SimulatedOS, "linux")
+	s.PatchValue(network.LaunchIpRouteShow, func() (string, error) {
+		return "default via 10.0.0.1 dev wlp2s0 proto static metric 800\n" +
+			"default via 10.100.1.10 metric 500\n" +
+			"10.0.0.0/24 dev wlp2s0 proto kernel scope link src 10.0.0.66 metric 600\n" +
+			"10.100.1.0/24 dev lxdbr0 proto kernel scope link src 10.100.1.1\n", nil
+	})
+	ip, dev, err := network.GetDefaultRoute()
+	c.Assert(err, gc.IsNil)
+	c.Check(ip, gc.DeepEquals, net.ParseIP("10.100.1.10"))
+	c.Check(dev, gc.Equals, "")
+}
+
+func (s *GatewaySuite) TestDefaultRouteLinuxError(c *gc.C) {
+	s.PatchValue(network.SimulatedOS, "linux")
+	s.PatchValue(network.LaunchIpRouteShow, func() (string, error) {
+		return "", fmt.Errorf("no can do")
+	})
+	ip, dev, err := network.GetDefaultRoute()
+	c.Check(ip, gc.IsNil)
+	c.Check(dev, gc.Equals, "")
+	c.Check(err, gc.ErrorMatches, "no can do")
+}
+
+func (s *GatewaySuite) TestDefaultRouteLinuxWrongOutput(c *gc.C) {
+	s.PatchValue(network.SimulatedOS, "linux")
+	s.PatchValue(network.LaunchIpRouteShow, func() (string, error) {
+		return "default via 10.0.0.1 dev wlp2s0 proto static metric chewbacca\n" +
+			"default dev lxdbr0\n" +
+			"10.0.0.0/24 dev wlp2s0 proto kernel scope link src 10.0.0.66 metric 600\n" +
+			"10.100.1.0/24 dev lxdbr0 proto kernel scope link src 10.100.1.1\n", nil
+	})
+	ip, dev, err := network.GetDefaultRoute()
+	c.Check(ip, gc.IsNil)
+	c.Check(dev, gc.Equals, "")
+	c.Check(err, gc.ErrorMatches, ".*chewbacca.*")
+}
+
+// We don't return 'not supported' error, we simply give no output
+func (s *GatewaySuite) TestDefaultRouteWindowsEmpty(c *gc.C) {
+	s.PatchValue(network.SimulatedOS, "windows")
+	s.PatchValue(network.LaunchIpRouteShow, func() (string, error) {
+		return "", fmt.Errorf("why someone calls me?")
+	})
+	ip, dev, err := network.GetDefaultRoute()
+	c.Check(ip, gc.IsNil)
+	c.Check(dev, gc.Equals, "")
+	c.Check(err, gc.IsNil)
+}
```

Comment:
The parsing side seems to be well tested.
Can we do some testing if we get malformed output from 'ip route' ? that we bail when things don't look correct rather than passing up bad data?
We should probably log at TRACE level the raw data that we're getting and processing.

Reply:
That's what we're doing in TestDefaultRouteLinuxWrongOutput, the output format is flexible and I don't think it's worth the effort to verify it strictly. If we see a proper 'default' line we will parse it properly, we will ignore the garbage.

Reply:
So that seems to test what happens if we fail to launch "ip route show" but not if we get back a line that we don't understand.